crates/ide-assists/src/handlers/recursively_add_derive.rs

```diff
@@ -0,0 +1,1010 @@
+                        make::token_tree(T!['('], tt).clone_for_update(),
+                    ))
+                    .clone_for_update();
+
+                    adt_src = edit.make_mut(adt_src);
+                    adt_src.add_attr(derive);
+                }
+            }
+        },
+    )
+}
+
+#[derive(Debug, Clone, Copy)]
+struct DeriveItem {
+    derive_macro: hir::Macro,
+    /// A guess at which trait the derive macro implements. This is used to check if a new derive needs to be added to the types of fields.
+    /// If [`None`], the derive is inserted if the type does not already contain it. This means manual trait implementations are ignored.
+    maybe_trait: Option<hir::Trait>,
+}
+
+impl DeriveItem {
+    fn from_path(
+        ctx: &AssistContext<'_>,
+        current_crate: hir::Crate,
+        ty: &hir::Type,
+        derive_macro: hir::Macro,
+        name: String,
+    ) -> Option<Self> {
+        if derive_macro.kind(ctx.db()) != hir::MacroKind::Derive {
+            return None;
+        }
+
+        // Try to find a trait with the same name as the derive macro, which the type implements.
+        let maybe_trait = items_locator::items_with_name(
+            &ctx.sema,
+            current_crate,
+            NameToImport::exact_case_sensitive(name),
+            items_locator::AssocSearchMode::Exclude,
+        )
+        .find_map(|item| item.as_module_def()?.as_trait())
+        .filter(|trait_| ty.impls_trait(ctx.db(), *trait_, &[]));
+        Some(DeriveItem { derive_macro, maybe_trait })
+    }
+
+    fn implemented_by(&self, ctx: &AssistContext<'_>, adt: &hir::Adt, src: &ast::Adt) -> bool {
+        // First check if the type already has the derive attribute
+        adt.attrs(ctx.db())
+            .by_key(&hir::sym::derive)
+            .attrs()
+            .filter_map(|attr| attr.parse_path_comma_token_tree(ctx.db()))
+            .flatten()
+            .filter_map(|(path, _)| ctx.sema.resolve_mod_path(src.syntax(), &path)?.find_map(|item| item.as_macro()))
+            .any(|macro_| self.derive_macro == macro_)
+            // If it does not, and there is a trait which the derive (likely) implements, check if the type has already implemented it.
+            || matches!(self.maybe_trait, Some(trait_) if adt.ty(ctx.db()).impls_trait(ctx.db(), trait_, &[]))
+    }
+}
+
+impl From<DeriveItem> for ItemInNs {
+    fn from(value: DeriveItem) -> Self {
+        value.derive_macro.into()
+    }
+}
+
+fn field_list_adts<'a>(
+    ctx: &'a AssistContext<'_>,
+    list: ast::FieldList,
+) -> impl Iterator<Item = hir::Adt> + 'a {
+    // The Option indirection here is to return a single iterator type for both types of field lists.
+    let mut record = None;
+    let mut tuple = None;
+    match list {
+        ast::FieldList::RecordFieldList(list) => {
+            record = Some(
+                list.fields().filter_map(|field| ctx.sema.resolve_type(&field.ty()?)?.as_adt()),
```

Comment:
It seems unfortunate to exclude things only because they are not ADT or not local, e.g. references will be excluded, and `Box`, and `Vec`, while all may have a conditional implementation for arbitrary types.

A better strategy will be when the type is non-local, to instead include its generic parameters (or their generic parameters, recursively). An even better (but harder) strategy will be to look if the type has a conditional implementation and what are its conditions (this doesn't have to go in the initial implementation, though).

Reply:
I completely forgot about that, I'll look into it 👍